release-note-script/src/main/java/ReleaseNoteCreation.java

```diff
@@ -0,0 +1,331 @@
+
+  private static final String DEBUG = System.getenv("DEBUG");
+
+  private final GitHubContext ghContext;
+
+  final Map<Category, List<ReleaseNoteText>> categoryMap = new EnumMap<>(Category.class);
+  final Map<String, List<ReleaseNoteText>> sameAsItems = new HashMap<>();
+
+  public static void main(String... args) throws Exception {
+    if (args.length != 4) {
+      System.err.printf(
+          "Usage:%n    java --source 11 %s.java <owner> <projectTitlePrefix> <version>"
+              + " <repository>%n%nExample:%n    java --source 11 %s.java scalar-labs ScalarDB 4.0.0"
+              + " scalardb%n",
+          ReleaseNoteCreation.class.getSimpleName(), ReleaseNoteCreation.class.getSimpleName());
+      System.exit(1);
+    }
+
+    String owner = args[0];
+    String projectTitlePrefix = args[1];
+    String version = args[2];
+    String repository = args[3];
+
+    ReleaseNoteCreation main =
+        new ReleaseNoteCreation(owner, projectTitlePrefix, version, repository);
+    main.createReleaseNote();
+  }
+
+  public ReleaseNoteCreation(
+      String owner, String projectTitlePrefix, String version, String repository) {
+    ghContext = new GitHubContext(owner, projectTitlePrefix, version, repository);
+  }
+
+  /** This constructor is only for test */
+  public ReleaseNoteCreation(GitHubContext ghContext) {
+    this.ghContext = ghContext;
+  }
+
+  public void createReleaseNote() throws Exception {
+    String projectId = ghContext.getProjectId();
+    List<String> prNumbers = ghContext.getPullRequestNumbers(projectId);
+
+    for (String prNumber : prNumbers) {
+      extractReleaseNoteInfo(prNumber);
+    }
+    assortSameAsItems();
+    outputReleaseNote();
+  }
+
+  void extractReleaseNoteInfo(String prNumber) throws Exception {
+    if (!ghContext.isPullRequestMerged(prNumber)) return;
+
+    ReleaseNoteText releaseNoteText = new ReleaseNoteText();
+    Category category = ghContext.getCategoryFromPullRequest(prNumber);
+
+    BufferedReader br = ghContext.getPullRequestBody(prNumber);
+
+    String line;
+    while ((line = br.readLine()) != null) {
+      if (Pattern.matches("^## *[Rr]elease *[Nn]otes? *", line)) {
+        releaseNoteText.category = category;
+        releaseNoteText.prNumbers.add(prNumber);
+
+        while ((line = br.readLine()) != null) {
+          if (Pattern.matches("^## *.*", line))
+            break; // Reached to the next section header (ended release note section)
+          if (Pattern.matches("^ *-? *N/?A *$", line)) return; // This PR is not user-facing
+
+          Matcher m =
+              Pattern.compile("^ *-? *(\\p{Print}+)$").matcher(line); // Extract Release note text
+          if (m.matches()) {
+            if (!Pattern.matches("^ *-? *[Ss]ame ?[Aa]s +#?([0-9]+) *$", line)) {
+              String matched = m.group(1);
+              if (DEBUG != null) System.err.printf("matched: %s%n", matched);
+              releaseNoteText.text = m.group(1);
+            }
+          }
```

Comment:
Since the logic of this method is complicated, I'd recommend extracting the sub-logic, like here, into separate methods.

Reply:
I agree. I'll revise it. Thank you.

Reply:
Thank you for the advice. But for now, I'd like to keep the code as is because I think it works without any problem. And the script is used only for release note creation workflow.

I will revise it according to your advice in another PR when I get a chance to revise it (e.g. Kaizen-day).